Backend/Schedules/Service/AppointmentService.cs

```diff
@@ -1,156 +1,91 @@
-                return appointment;
-            }
+            List<Appointment> available = GetAvailableByDoctorAndDate(appointment.DoctorId, appointment.Start);
+            bool isAvailable = available.Any(a => a.isOccupied(appointment.Start, appointment.End));
+            if (isAvailable)
+                return _appointmentRepository.Create(appointment);
+
             return null;
         }
-        public Appointment ChangeRoomForAppointment(Appointment appointment, Room room)
+        public List<Appointment> GetAvailableByDoctorAndDateRange(string doctorId, DateTime start, DateTime end)
         {
-            Appointment appointmentToUpdate = appointmentRepository.GetObject(appointment.Id);
-            appointmentToUpdate.Room = room;
-            return appointmentRepository.Update(appointmentToUpdate);
+            List<Appointment> availableAppointments = new List<Appointment>();
+            for (DateTime date = start; date.Date <= end.Date; date = date.AddDays(1))
+            {
+                availableAppointments.AddRange(GetAvailableByDoctorAndDate(doctorId, date));
+            }
+            return availableAppointments;
         }
-        public IEnumerable<Appointment> NotFinishedByDoctorAndDay(Doctor doctor, DateTime day)
+        public List<Appointment> GetAvailableByDoctorAndDate(string doctorId, DateTime date)
         {
-            var doctorsDay = GetScheduledBy(doctor, day);
-            return doctorsDay.Where(entity => !entity.Finished);
+            List<Appointment> occupied = GetByDoctorAndDate(doctorId, date);
+            List<Appointment> allAppointments = InitializeAppointments(doctorId, date);
+            List<Appointment> available = new List<Appointment>(allAppointments);
+            foreach(Appointment appointmentIt in allAppointments)
+            {
+                Appointment appointment = occupied.FirstOrDefault(a => a.isOccupied(appointmentIt.Start, appointmentIt.End));
+                if (isOccupied(appointment))
+                    available.Remove(appointmentIt);
+            }
+            return available;
         }
-        public int GetNumberOfAppointmentsFor(Doctor doctor, TypeOfAppointment type)
+        private bool isOccupied(Appointment appointment)
         {
-            var allExams = appointmentRepository.GetAll().ToList().Where(entity => entity.IsDoctor(doctor));
-            return allExams.Where(entity => entity.Finished == true && entity.TypeOfAppointment == type).ToList().Count;
+            return appointment != null && !appointment.canceledByPatient;
         }
-
-        public Appointment FinishAppointment(Appointment appointment)
+    
+        public List<Appointment> GetByDoctorAndDate(string doctorId, DateTime date)
         {
-            appointment.Finished = true;
-            return appointmentRepository.Update(appointment);
+            return _appointmentRepository.GetByDoctorIdAndDate(doctorId, date).ToList();
         }
 
```

Comment:
Also, you can remove _DoctorAndDate_, and name the method GetBy or if you want to be more precise, GetScheduledBy.